### PR TITLE
[enhancement](Nereids): add more implmentation rules for ssb.

### PR DESCRIPTION
**Pull request by jackwener**
# Proposed changes

Issue Number: close #xxx

## Problem Summary:

Add more implmentation rules.

Current some `logical` and `physical` operator is different. I change some code to make them match.

Implementation
- Sort
- Agg
- OlapScan

## Checklist(Required)

1. Does it affect the original behavior: (No)
2. Has unit tests been added: (No Need)
3. Has document been added or modified: (No Need)
4. Does it need to update dependencies: (No)
5. Are there any changes that cannot be rolled back: (No)

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/implementation/LogicalProjectionToPhysicalProjection.java [hunk 1]
```diff
@@ -23,7 +23,7 @@
 import org.apache.doris.nereids.trees.plans.Plan;
 
 /**
- * Implementation rule that convert logical join to physical hash join.
+ * Implementation rule that convert logical projection to physical projection.
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/rules/implementation/LogicalProjectionToPhysicalProjection.java:26]:
```suggestion
 * Implementation rule that convert logical project to physical project.
```

**Comment by morrySnow** [comment]
Could you list what implementation rules add in this PR in '**Proposed changes**'

File: fe/fe-core/src/main/java/org/apache/doris/nereids/operators/plans/logical/LogicalOlapScan.java [hunk 2]
```diff
@@ -0,0 +1,21 @@
+package org.apache.doris.nereids.operators.plans.logical;
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/operators/plans/logical/LogicalOlapScan.java:1]: header

File: fe/fe-core/src/main/java/org/apache/doris/nereids/operators/plans/physical/PhysicalSort.java [hunk 3]
```diff
@@ -33,22 +33,19 @@
...
 public class PhysicalSort extends PhysicalUnaryOperator {
-
+    // Default offset is 0.
     private final int offset;
 
-    private final int limit;
-
     private final List<OrderKey> orderList;
-
+    // TODO(wj): Do we need it? If offset is 0 and limit != 0, we can think it's topN.
```
[the remark this answers is not in the document]
Comment (jackwener) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/operators/plans/physical/PhysicalSort.java:40]: It's is a question worth considering.

**Comment by EmmyMiao87** [comment]
It is recommended to change the useless List<? extends A> in the code to List<A>

File: fe/fe-core/src/main/java/org/apache/doris/nereids/operators/plans/logical/LogicalAggregation.java [hunk 4]
```diff
@@ -31,68 +32,70 @@
...
+
+    public AggPhase getAggPhase() {
+        return aggPhase;
     }
 
     @Override
     public String toString() {
-        return "Aggregation (" + "outputExpressions: " + StringUtils.join(outputExpressions, ", ")
-                + ", groupByExpressions: " + StringUtils.join(groupByExpressions, ", ") + ")";
+        return "Aggregation (" + "outputExpressions: " + StringUtils.join(aggExprList, ", ") + ", groupByExpressions: "
```
Comment (yinzhijian) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/operators/plans/logical/LogicalAggregation.java:82]: Is it better to unify it with the variable name here？outputExpressions=>aggExprList, etc.

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/implementation/LogicalAggToHashAgg.java [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+public class LogicalAggToHashAgg extends OneImplementationRuleFactory {
+    @Override
+    public Rule<Plan> build() {
+        return logicalAggregation().then(agg -> plan(
+            new PhysicalAggregation(
+                agg.getOperator().getGroupByExprList(),
+                agg.getOperator().getAggExprList(),
+                agg.getOperator().getPartitionExprList(),
+                agg.getOperator().getAggPhase(),
+                false),
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/rules/implementation/LogicalAggToHashAgg.java:37]: add a todo for use a function to judge whether use stream

File: fe/fe-core/src/main/java/org/apache/doris/nereids/operators/plans/physical/PhysicalHeapSort.java [hunk 6]
```diff
@@ -32,37 +32,29 @@
...
+public class PhysicalHeapSort extends PhysicalUnaryOperator {
+    // Default offset is 0.
     private final int offset;
 
-    private final int limit;
-
-    private final List<OrderKey> orderList;
-
+    private final List<OrderKey> orderKeys;
     private final boolean useTopN;
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/operators/plans/physical/PhysicalHeapSort.java:40]: useTopN need to be removed. In translator, HeapSort will translate to topN and MergeSort will translate to normal sort

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/implementation/LogicalOlapScanToPhysicalOlapScan.java [hunk 7]
```diff
@@ -0,0 +1,37 @@
...
+import org.apache.doris.nereids.trees.plans.Plan;
+
+/**
+ * Implementation rule that convert logical OlapScan to physical OlapScan.
+ */
+public class LogicalOlapScanToPhysicalOlapScan extends OneImplementationRuleFactory {
+    @Override
+    public Rule<Plan> build() {
+        return logicalJoin().then(join -> plan(
+                new PhysicalHashJoin(join.operator.getJoinType(), join.operator.getCondition().get()),
```
Comment (yinzhijian) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/rules/implementation/LogicalOlapScanToPhysicalOlapScan.java:32]: PhysicalHashJoin?

File: fe/fe-core/src/main/java/org/apache/doris/nereids/operators/plans/logical/LogicalAggregation.java [hunk 8]
```diff
@@ -31,68 +32,70 @@
...
 
-    /**
-     * Get GroupByAggregation list.
-     *
-     * @return all group by of this node.
-     */
-    public List<Expression> getGroupByExpressions() {
-        return groupByExpressions;
+    public List<Expression> getPartitionExprList() {
+        return partitionExprList;
```
Comment (EmmyMiao87) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/operators/plans/logical/LogicalAggregation.java:61]:
```suggestion
        return partitionExprList == null ? groupByExprList : partitionExprList;
```

File: fe/fe-core/src/main/java/org/apache/doris/nereids/operators/plans/physical/PhysicalAggregation.java [hunk 9]
```diff
@@ -88,10 +89,7 @@ public <R, C> R accept(PlanOperatorVisitor<R, C> visitor, Plan plan, C context)
 
     @Override
     public List<Expression> getExpressions() {
-        return new ImmutableList.Builder<Expression>()
-                .addAll(groupByExprList)
-                .addAll(aggExprList)
-                .addAll(partitionExprList)
-                .build();
+        return new ImmutableList.Builder<Expression>().addAll(groupByExprList).addAll(outputExpressionList)
+                .addAll(partitionExprList).build();
```
Comment (EmmyMiao87) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/operators/plans/physical/PhysicalAggregation.java:93]: partitionExprList maybe null

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/implementation/LogicalAggToHashAgg.java [hunk 10]
```diff
@@ -0,0 +1,43 @@
...
+
+import org.apache.doris.nereids.operators.plans.physical.PhysicalAggregation;
+import org.apache.doris.nereids.rules.Rule;
+import org.apache.doris.nereids.rules.RuleType;
+import org.apache.doris.nereids.trees.plans.Plan;
+
+/**
+ * Implementation rule that convert logical aggregation to physical hash aggregation.
+ */
+public class LogicalAggToHashAgg extends OneImplementationRuleFactory {
```
Comment (EmmyMiao87) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/rules/implementation/LogicalAggToHashAgg.java:28]:
```suggestion
public class LogicalAggToPhysicalHashAgg extends OneImplementationRuleFactory {
```

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/RuleType.java [hunk 11]
```diff
@@ -42,12 +42,12 @@ public enum RuleType {
...
     // implementation rules
+    LOGICAL_AGG_TO_HASH_AGG_RULE(RuleTypeClass.IMPLEMENTATION),
     LOGICAL_JOIN_TO_HASH_JOIN_RULE(RuleTypeClass.IMPLEMENTATION),
-    LOGICAL_PROJECTION_TO_PHYSICAL_PROJECTION_RULE(RuleTypeClass.IMPLEMENTATION),
+    LOGICAL_PROJECT_TO_PHYSICAL_PROJECT_RULE(RuleTypeClass.IMPLEMENTATION),
     LOGICAL_FILTER_TO_PHYSICAL_FILTER_RULE(RuleTypeClass.IMPLEMENTATION),
-    LOGICAL_LIMIT_TO_PHYSICAL_LIMIT_RULE(RuleTypeClass.IMPLEMENTATION),
-    LOGICAL_JOIN_TO_HASH_AGG_RULE(RuleTypeClass.IMPLEMENTATION),
-    LOGICAL_JOIN_TO_OLAP_SCAN_RULE(RuleTypeClass.IMPLEMENTATION),
+    LOGICAL_HEAP_SORT_TO_PHYSICAL_HEAP_SORT_RULE(RuleTypeClass.IMPLEMENTATION),
```
Comment (EmmyMiao87) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/rules/RuleType.java:49]:
```suggestion
    LOGICAL_SORT_TO_PHYSICAL_HEAP_SORT_RULE(RuleTypeClass.IMPLEMENTATION),
```

File: fe/fe-core/src/main/java/org/apache/doris/nereids/operators/plans/logical/LogicalAggregation.java [hunk 12]
```diff
@@ -31,68 +32,70 @@
 
 /**
  * Logical Aggregation plan operator.
- *
- *eg:select a, sum(b), c from table group by a, c;
+ * <p>
+ * eg:select a, sum(b), c from table group by a, c;
  * groupByExpressions: Column field after group by. eg: a, c;
  * outputExpressions: Column field after select. eg: a, sum(b), c;
```
[the remark this answers is not in the document]
Comment (morrySnow) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/operators/plans/logical/LogicalAggregation.java:38]: i think the `partitionExprList` just copy from stale planner, may be we do not need it anymore in future

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/RuleType.java (hunk 11 above)
Comment (morrySnow) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/rules/RuleType.java:49]: we do not add the 'Physical' prefix if we have more than one physical implementation for one operator

File: fe/fe-core/src/main/java/org/apache/doris/nereids/trees/plans/PhysicalPlanTranslator.java [hunk 13]
```diff
@@ -147,20 +148,20 @@ public PlanFragment visitPhysicalOlapScan(
     }
 
     @Override
-    public PlanFragment visitPhysicalSort(PhysicalUnaryPlan<PhysicalSort, Plan> sort,
+    public PlanFragment visitPhysicalSort(PhysicalUnaryPlan<PhysicalHeapSort, Plan> sort,
             PlanTranslatorContext context) {
         PlanFragment childFragment = visit(sort.child(0), context);
-        PhysicalSort physicalSort = sort.getOperator();
+        PhysicalHeapSort physicalHeapSort = sort.getOperator();
         if (!childFragment.isPartitioned()) {
```
[the remark this answers is not in the document]
Comment (morrySnow) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/trees/plans/PhysicalPlanTranslator.java:155]: it is a bug introduced by @Kikyou1997, please fix it~

File: fe/fe-core/src/main/java/org/apache/doris/nereids/operators/plans/logical/LogicalAggregation.java (hunk 12 above)
[the remark this answers is not in the document]
Comment (EmmyMiao87) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/operators/plans/logical/LogicalAggregation.java:38]: I agree with you ~

**Review by EmmyMiao87 — APPROVED** [review]
LGTM

**Pull request merged**